JDBI throws MismatchedTokenException(-1!=39) encountering a quote (') inside comment

Affects: JDBI 2.49

The culprit is: `/* in warehouse's time zone */` in the query, changing it to `/* in warehouse time zone */` makes JDBI happy.

Stacktrace:

```
Caused by: org.skife.jdbi.v2.exceptions.CallbackFailedException: org.skife.jdbi.v2.exceptions.UnableToCreateStatementException: Exception parsing for named parameter replacement [statement:"null", located:"null", rewritten:"null", arguments:null]
	org.skife.jdbi.v2.DBI.withHandle(DBI.java:277) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.DBI.inTransaction(DBI.java:299) ~[jdbi-2.49.jar:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository.createQtyTable(JournalPerpetualInventoryRepository.java:220) ~[classes/:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository.createMissingTables(JournalPerpetualInventoryRepository.java:256) ~[classes/:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository.initTables(JournalPerpetualInventoryRepository.java:196) ~[classes/:na]
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)(NativeMethodAccessorImpl.java) ~[na:1.7.0_25]
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_25]
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_25]
	java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_25]
	org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:344) ~[spring-beans-3.2.3.RELEASE.jar:3.2.3.RELEASE]
	org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:295) ~[spring-beans-3.2.3.RELEASE.jar:3.2.3.RELEASE]
	org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:130) ~[spring-beans-3.2.3.RELEASE.jar:3.2.3.RELEASE]
	... 65 more
Caused by: org.skife.jdbi.v2.exceptions.UnableToCreateStatementException: Exception parsing for named parameter replacement [statement:"null", located:"null", rewritten:"null", arguments:null]
	org.skife.jdbi.v2.ColonPrefixNamedParamStatementRewriter.rewrite(ColonPrefixNamedParamStatementRewriter.java:60) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.Batch.execute(Batch.java:107) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.Script.execute(Script.java:72) ~[jdbi-2.49.jar:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository.executeScript(JournalPerpetualInventoryRepository.java:288) ~[classes/:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository.access$1(JournalPerpetualInventoryRepository.java:282) ~[classes/:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository$1.inTransaction(JournalPerpetualInventoryRepository.java:233) ~[classes/:na]
	id.co.bippo.inventory.perpetual.JournalPerpetualInventoryRepository$1.inTransaction(JournalPerpetualInventoryRepository.java:1) ~[classes/:na]
	org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:193) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:309) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.DBI$4.withHandle(DBI.java:302) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.DBI.withHandle(DBI.java:274) ~[jdbi-2.49.jar:na]
	... 76 more
Caused by: java.lang.IllegalArgumentException: MismatchedTokenException(-1!=39)
	org.skife.jdbi.rewriter.colon.ColonStatementLexer.reportError(ColonStatementLexer.java:24) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.org.antlr.runtime.Lexer.nextToken(Lexer.java:98) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.ColonPrefixNamedParamStatementRewriter.parseString(ColonPrefixNamedParamStatementRewriter.java:93) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.v2.ColonPrefixNamedParamStatementRewriter.rewrite(ColonPrefixNamedParamStatementRewriter.java:56) ~[jdbi-2.49.jar:na]
	... 86 more
Caused by: org.skife.jdbi.org.antlr.runtime.MismatchedTokenException
	org.skife.jdbi.org.antlr.runtime.Lexer.match(Lexer.java:188) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.rewriter.colon.ColonStatementLexer.mQUOTED_TEXT(ColonStatementLexer.java:269) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.rewriter.colon.ColonStatementLexer.mTokens(ColonStatementLexer.java:557) ~[jdbi-2.49.jar:na]
	org.skife.jdbi.org.antlr.runtime.Lexer.nextToken(Lexer.java:84) ~[jdbi-2.49.jar:na]
	... 88 more
```

Complete query as follows:

```
CREATE TABLE jnlinv_qty (
	qty_id bigserial primary key,
	/* Built-in dimensions */
	movement_time timestamptz NOT NULL,
	warehouse_id varchar(255) NOT NULL REFERENCES jnlinv_warehouse(warehouse_id),
	storage_bin_id varchar(255) NOT NULL REFERENCES jnlinv_storage_bin(storage_bin_id),
	product_id varchar(255) NOT NULL REFERENCES jnlinv_product(product_id),
	/* Principals */
	base_color varchar(255) NOT NULL REFERENCES jnlinv_base_color(base_color),
	base_size varchar(255) NOT NULL REFERENCES jnlinv_base_size(base_size),

	/* Measures */
	qty_on_hand numeric(24, 8) NOT NULL,
	qty_reserved numeric(24, 8) NOT NULL,
	qty_picked numeric(24, 8) NOT NULL,
	qty_available numeric(24, 8) NOT NULL,
	qty_produced numeric(24, 8) NOT NULL,
	/* Convenience for mutation report */
	qty_on_hand_delta numeric(24,8),         
	qty_on_hand_update numeric(24,8),        
	qty_reserved_delta numeric(24,8),        
	qty_picked_delta numeric(24,8),          
	/* Informational columns */
	physical_inventory_id varchar(36),
	stock_reservation_id varchar(36),
	goods_receipt_id varchar(36),
	return_to_vendor_shipment_id varchar(36),
	goods_shipment_id varchar(36),
	return_material_receipt_id varchar(36),
	movement_local_time timestamp NOT NULL, /* in warehouse's time zone */
	/* Administrative columns */
	creation_time timestamptz NOT NULL default 'now()',
	/* Key (combined dimensions) must be unique */
	CONSTRAINT jnlinv_dimensions_key UNIQUE (movement_time, warehouse_id, storage_bin_id, product_id, 
                                                                                                   base_color, 
                                                                                                   base_size)
);

CREATE INDEX idx_jnlinv_qty_movement_time_id ON jnlinv_qty (movement_time);
CREATE INDEX idx_jnlinv_qty_movement_local_time_id ON jnlinv_qty (movement_local_time);
/* Built-in dimensions */
CREATE INDEX idx_jnlinv_qty_warehouse_id ON jnlinv_qty (warehouse_id);
CREATE INDEX idx_jnlinv_qty_storage_bin_id ON jnlinv_qty (storage_bin_id);
CREATE INDEX idx_jnlinv_qty_product_id ON jnlinv_qty (product_id);
/* Principals */
CREATE INDEX idx_jnlinv_qty_base_color ON jnlinv_qty (base_color);
CREATE INDEX idx_jnlinv_qty_base_size ON jnlinv_qty (base_size);
```

--- a/src/test/java/org/skife/jdbi/v2/TestColonStatementRewriter.java
+++ b/src/test/java/org/skife/jdbi/v2/TestColonStatementRewriter.java
@@ -124,4 +124,8 @@ public class TestColonStatementRewriter
         assertEquals(1, ctr.get());
     }
 
+    public void testCommentQuote() throws Exception
+    {
+        rw.rewrite("select 1 /* ' \" */", new Binding(), new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+    }
 }

--- a/src/test/java/org/skife/jdbi/v2/TestColonStatementRewriter.java
+++ b/src/test/java/org/skife/jdbi/v2/TestColonStatementRewriter.java
@@ -18,16 +18,12 @@ import org.junit.Before;
 import org.junit.Test;
 import org.skife.jdbi.v2.exceptions.UnableToCreateStatementException;
 import org.skife.jdbi.v2.tweak.RewrittenStatement;
-import org.skife.jdbi.v2.tweak.StatementRewriter;
 
 import java.util.HashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.junit.Assert.assertEquals;
 
-/**
- *
- */
 public class TestColonStatementRewriter
 {
     private ColonPrefixNamedParamStatementRewriter rw;
@@ -38,72 +34,67 @@ public class TestColonStatementRewriter
         this.rw = new ColonPrefixNamedParamStatementRewriter();
     }
 
+    private RewrittenStatement rewrite(String sql)
+    {
+        return rw.rewrite(sql,
+                new Binding(),
+                new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+    }
+
     @Test
     public void testNewlinesOkay() throws Exception
     {
-        RewrittenStatement rws = rw.rewrite("select * from something\n where id = :id", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        RewrittenStatement rws = rewrite("select * from something\n where id = :id");
         assertEquals("select * from something\n where id = ?", rws.getSql());
     }
 
     @Test
     public void testOddCharacters() throws Exception
     {
-        RewrittenStatement rws = rw.rewrite("~* :boo ':nope' _%&^& *@ :id", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        RewrittenStatement rws = rewrite("~* :boo ':nope' _%&^& *@ :id");
         assertEquals("~* ? ':nope' _%&^& *@ ?", rws.getSql());
     }
 
     @Test
     public void testNumbers() throws Exception
     {
-        RewrittenStatement rws = rw.rewrite(":bo0 ':nope' _%&^& *@ :id", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        RewrittenStatement rws = rewrite(":bo0 ':nope' _%&^& *@ :id");
         assertEquals("? ':nope' _%&^& *@ ?", rws.getSql());
     }
 
     @Test
     public void testDollarSignOkay() throws Exception
     {
-        RewrittenStatement rws = rw.rewrite("select * from v$session", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        RewrittenStatement rws = rewrite("select * from v$session");
         assertEquals("select * from v$session", rws.getSql());
     }
 
     @Test
     public void testHashInColumnNameOkay() throws Exception
     {
-       RewrittenStatement rws = rw.rewrite("select column# from thetable where id = :id", new Binding(),
-                                           new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+       RewrittenStatement rws = rewrite("select column# from thetable where id = :id");
        assertEquals("select column# from thetable where id = ?", rws.getSql());
     }
 
     @Test
     public void testBacktickOkay() throws Exception
     {
-        RewrittenStatement rws = rw.rewrite("select * from `v$session", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        RewrittenStatement rws = rewrite("select * from `v$session");
         assertEquals("select * from `v$session", rws.getSql());
     }
 
-    @Test
+    @Test(expected = UnableToCreateStatementException.class)
     public void testBailsOutOnInvalidInput() throws Exception
     {
-        try {
-            rw.rewrite("select * from something\n where id = :\u0087\u008e\u0092\u0097\u009c", new Binding(),
-                                            new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
-
-            Assert.fail("Expected 'UnableToCreateStatementException' but got none");
-        }
-        catch (UnableToCreateStatementException e) {
-        }
+        rewrite("select * from something\n where id = :\u0087\u008e\u0092\u0097\u009c");
+        Assert.fail("Expected 'UnableToCreateStatementException' but got none");
     }
 
     @Test
     public void testCachesRewrittenStatements() throws Exception
     {
         final AtomicInteger ctr = new AtomicInteger(0);
-        ColonPrefixNamedParamStatementRewriter rw = new ColonPrefixNamedParamStatementRewriter()
+        rw = new ColonPrefixNamedParamStatementRewriter()
         {
             @Override
             ParsedStatement parseString(final String sql) throws IllegalArgumentException
@@ -113,19 +104,17 @@ public class TestColonStatementRewriter
             }
         };
 
-        rw.rewrite("insert into something (id, name) values (:id, :name)", new Binding(),
-                new ConcreteStatementContext(new HashMap<String, Object>(), null));
+        rewrite("insert into something (id, name) values (:id, :name)");
 
         assertEquals(1, ctr.get());
 
-        rw.rewrite("insert into something (id, name) values (:id, :name)", new Binding(),
-                new ConcreteStatementContext(new HashMap<String, Object>(), null));
+        rewrite("insert into something (id, name) values (:id, :name)");
 
         assertEquals(1, ctr.get());
     }
 
     public void testCommentQuote() throws Exception
     {
-        rw.rewrite("select 1 /* ' \" */", new Binding(), new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry()));
+        rewrite("select 1 /* ' \" */");
     }
 }

